File: python/chronopulser_decoder.py
import json

def load_cipher(path="../chronopulse.json"):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)

def decode_chronopulse(code, cipher):
    alphabet = cipher["Alphabet"]
    reverse = {v: k for k, v in alphabet.items()}
    symbols = cipher["Symbols"]

    result, current = [], ""

    for s in code:
        if s in ("□", "■"):
            current += s
        elif s == "●":
            if current:
                result.append(reverse.get(current, "?"))
                current = ""
        elif s == "○":
            result.append(" ")
        elif s == "•":
            result.append(", ")
            current = ""
        elif s in symbols:
            result.append(symbols[s]["Represents"])
        else:
            result.append("?")

    if current:
        result.append(reverse.get(current, "?"))

    return "".join(result).strip()

if __name__ == "__main__":
    cipher = load_cipher()
    code = input("Enter code to decode: ")
    print(decode_chronopulse(code, cipher))

File: lua/chronopulser_decoder.lua
local json = require("dkjson")

local function loadCipher(path)
    local f = assert(io.open(path, "r"))
    local text = f:read("*a")
    f:close()
    return json.decode(text)
end

local function decodeChronoPulse(code, cipher)
    local alphabet = cipher.Alphabet
    local reverse = {}
    for k,v in pairs(alphabet) do reverse[v] = k end
    local symbols = cipher.Symbols

    local result, current = {}, ""
    for i = 1, #code do
        local s = code:sub(i,i)
        if s=="□" or s=="■" then current = current..s
        elseif s=="●" then
            table.insert(result, reverse[current] or "?")
            current=""
        elseif s=="○" then table.insert(result," ")
        elseif s=="•" then table.insert(result,", "); current=""
        elseif symbols[s] then table.insert(result,symbols[s].Represents)
        else table.insert(result,"?")
        end
    end
    if current~="" then table.insert(result, reverse[current] or "?") end
    return table.concat(result)
end

local cipher = loadCipher("../chronopulse.json")
io.write("Enter code to decode: ")
local code = io.read()
print(decodeChronoPulse(code, cipher))
